Source/MaestroMakeEdgeScalars.cpp

```diff
@@ -0,0 +1,1100 @@
+    Real hx = dx[0];
+    Real hy = dx[1];
+    Real hz = dx[2];
+
+    // Get the index space of the valid region
+    const Box& tileBox = mfi.tilebox();
+    const Box& obx = amrex::grow(tileBox, 1);
+    const Box& mxbx = amrex::growLo(obx, 0, -1);
+    const Box& mybx = amrex::growLo(obx, 1, -1);
+    const Box& mzbx = amrex::growLo(obx, 2, -1);
+
+    // loop over appropriate x-faces
+    int ilo = domainBox.loVect()[0];
+    int ihi = domainBox.hiVect()[0];
+    int bclo = bcs[bccomp].lo()[0];
+    int bchi = bcs[bccomp].hi()[0];
+    AMREX_PARALLEL_FOR_3D(mxbx, i, j, k, 
+    {
+        if (ppm_type_local == 0) {
+            slx(i,j,k) = scal(i-1,j,k,comp) + 
+                0.5 * (1.0 - dt * umac(i,j,k) / hx) * Ip(i-1,j,k,0);
+            srx(i,j,k) = scal(i,j,k,comp) - 
+                0.5 * (1.0 + dt * umac(i,j,k) / hx) * Ip(i,j,k,0);
+        } else if (ppm_type_local == 1 || ppm_type_local == 2) {
+            slx(i,j,k) = Ip(i-1,j,k,0);
+            srx(i,j,k) = Im(i,j,k,0);
+        }
+
+        // impose lo side bc's
+        if (i == ilo) {
+            if (bclo == EXT_DIR) {
+                slx(i,j,k) = scal(i-1,j,k,comp);
+                srx(i,j,k) = scal(i-1,j,k,comp);
+            } else if (bclo == FOEXTRAP || bclo == HOEXTRAP) {
+                if (is_vel && comp == 0) {
+                    srx(i,j,k) = min(srx(i,j,k), 0.0);
+                }
+                slx(i,j,k) = srx(i,j,k);
+            } else if (bclo == REFLECT_EVEN) {
+                slx(i,j,k) = srx(i,j,k);
+            } else if (bclo == REFLECT_ODD) {
+                slx(i,j,k) = 0.0;
+                srx(i,j,k) = 0.0;
+            }
+
+        // impose hi side bc's
+        } else if (i == ihi+1) {
+            if (bchi == EXT_DIR) {
+                slx(i,j,k) = scal(i,j,k,comp);
+                srx(i,j,k) = scal(i,j,k,comp);
+            } else if (bchi == FOEXTRAP || bchi == HOEXTRAP) {
+                if (is_vel && comp == 0) {
+                    slx(i,j,k) = max(slx(i,j,k), 0.0);
+                }
+                srx(i,j,k) = slx(i,j,k);
+            } else if (bchi == REFLECT_EVEN) {
+                srx(i,j,k) = slx(i,j,k);
+            } else if (bchi == REFLECT_ODD) {
+                slx(i,j,k) = 0.0;
+                srx(i,j,k) = 0.0;
+            }
+        }
+
+        // make simhx by solving Riemann problem
+        simhx(i,j,k) = (umac(i,j,k) > 0.0) ? 
+            slx(i,j,k) : srx(i,j,k);
+        simhx(i,j,k) = (fabs(umac(i,j,k)) > 0.0) ? 
+            simhx(i,j,k) : 0.5 * (slx(i,j,k) + srx(i,j,k));
+    });
+
+    // loop over appropriate y-faces
+    int jlo = domainBox.loVect()[1];
+    int jhi = domainBox.hiVect()[1];
+    bclo = bcs[bccomp].lo()[1];
+    bchi = bcs[bccomp].hi()[1];
+    AMREX_PARALLEL_FOR_3D(mybx, i, j, k, 
+    {
+        if (ppm_type_local == 0) {
+            sly(i,j,k) = scal(i,j-1,k,comp) + 
+                0.5 * (1.0 - dt * vmac(i,j,k) / hy) * Ip(i,j-1,k,0);
```

Comment:
should these slopes have the last index of `1` instead of `0`?

Reply:
or be `Im`?

Reply:
no, but they should be Im not Ip. For ppm type 0, Ip stores slopex and Im stores slopey